File: day1/main.py

from collections import defaultdict


if __name__ == "__main__":

    fn = "input.txt"

    with open(fn, "r") as file:
        lines = [line.strip() for line in file.readlines()]

    lines = [[int(n) for n in line.split("   ")] for line in lines]

    left, right = list(zip(*lines))
    left = list(left)
    right = list(right)

    left.sort()
    right.sort()

    diff = [abs(left[i] - right[i]) for i in range(len(left))]

    print(sum(diff))

    right_count = defaultdict(int)
    for r in right:
        right_count[r] += 1


    cumulative = 0
    for l in left:
        cumulative += l * right_count[l]

    print(cumulative)

File: day1/main.cpp

#include <iostream>
#include <fstream>
#include <vector>

int main() {
    std::ifstream file("input.txt");
    std::vector<int> left;
    std::vector<int> right;
    int l, r;
    while (file >> l >> r) {
        left.push_back(l);
        right.push_back(r);
    }

    std::sort(left.begin(), left.end());
    std::sort(right.begin(), right.end());

    int diff = 0;
    for (int i = 0; i < left.size(); i++) {
        diff += abs(left[i] - right[i]);
    }
    std::cout << "Total difference: " << diff << std::endl;

    // Part 2
    std::unordered_map<int, int> freq;
    for (int n : right) {
        freq[n]++;
    }

    long long sum = 0;
    for (int n : left) {
        sum += n * freq[n];
    }
    std::cout << "Total sum: " << sum << std::endl;
}